src/library/scala/sys/process/ProcessBuilderImpl.scala

```diff
@@ -170,11 +170,14 @@ private[process] trait ProcessBuilderImpl {
       val lazilyListed = LazilyListed[String](nonZeroException, capacity)
       val process      = run(BasicIO(withInput, lazilyListed.process, log))
 
+      // extract done from lazilyListed so that the anonymous function below closes over just the done and not the whole lazilyListed (see https://github.com/scala/bug/issues/12185)
+      val done = lazilyListed.done
+
       Spawn("LazyLines") {
-        lazilyListed.done {
+        done {
           try process.exitValue()
           catch {
-            case NonFatal(_) => -2
+            case _: Exception => -2
```

Comment:
I’m not sure about this change. `NonFatal` catches `StackoverflowError`, for instance, which is not anymore caught by your change.

Reply:
OOME is not nonfatal.
```
scala> try { throw new OutOfMemoryError } catch { case NonFatal(_) => println("whew") }
java.lang.OutOfMemoryError
  at liftedTree1$1(<console>:1)
  ... 32 elided
```

Reply:
@julienrf are you sure about that StackoverflowError? I think that changed in some version of Scala and now it doesn't.
@som-snytt I know `OutOfMemoryError` is not `NonFatal`. I don't fully understand what is happening here, but what I observed was that with the NonFatal the JVM kept running GCs infinitely and not printing OOME. With the `_: Exception` I ended up getting the error printed with each run. The problem is I can't reproduce it now, except for the one JVM in the infinite GC state that I haven't restarted yet.

Caching `Error`s is a tricky business, and I should have been running my JVMs with `-XX:+CrashOnOutOfMemoryError` anyway. I would be ok with reverting the change to the `catch`, and proceeding just with the `val done = lazilyListed.done` part, which is IMO the important part of this PR. @julienrf @som-snytt would you like to proceed this way?

Reply:
yes, the other change is a good catch. It took me a few a minutes to get it. I'll try to come up with a test, which would be a mock Process that supplies some data and then (I guess) checking in the middle of the fold that the first "line" is not reachable (from whatever references we know about). We have other unit tests that chase references; it is a bit tricky.

On the OOM, I think there is a threshold you can set for how much time to spend in gc before giving up. Maybe you expected it to fail faster?

Reply:
Ok. I've remove the commit modifying the catch.

It's possible I'm oscillating near the threshold for GC overhead OOME.